### DISCUSSION TITLE
Package name alias

### DISCUSSION THREAD

**Discussion by ghost** [composer/composer#11851] [Q&A / Support]
**The problem**
We have a package that is used by many teams. We did some refactoring and there are better names that describe what the package is for than the current name. We don't have space to force all the teams to update right now but my team would like to use new name. Also changing the main name will probably break some CI and that would require time to adjust.

**The solution**
I would like to give my package two names. Let's say I have `vendor/package-foo` a the moment. I would like to have second name `vendor/package-foo-alias`. Then `composer require vendor/package-foo` and `composer require vendor/package-foo-alias` would install the same package.
I imagine it could look something like that in composer.json:
```json
{
  "name" : "vendor/package-foo",
  "name-alias": "vendor/package-foo-alias"
  // or
  "name-aliases": [
    "vendor/package-foo-alias",
    "vendor/package-foo-another-alias"
  ]
}
```

Is it possible at the current state of Composer?

**Comment by fredden** [comment]
A metapackage may help with this transition. You could rename your package, and publish a metapackage at the old name, which simply requires (an) appropriate version(s) of the new package name. Maybe also set an `abandoned` field on the metapackage pointing to the new package, so that users get some notification about the underlying change.

**Reply by ghost** [reply]
I thought about such a solution but for the new name - then the old repository would stay as is (links, CI). However this is not a satisfying solution because this is another repository to maintain. Also how would I manage tags when someone requires the package by new name at the given tag.
Having an package name alias would solve the issue in one line.

**Comment by stof** [comment]
You can use `provide` or `replace` as a way to make your new name satisfy requirements for the old package name (in the case of a package renaming, `replace` is more appropriate, as `provide` is meant for virtual packages).

However, you would still need to have one place in your dependency graph requiring the actual package name and not only the aliased name (this was not true in early composer version, but the restriction was put in place based on security rules to prevent arbitrary package highjacking).

**Reply by ghost** [reply]
If I understand `replace` correctly then it will not work for me. I would like to change the package name in the BC manner as the whole company cannot migrate at the same time. IIRC Composer is looking for the package name on the default branch (we use Satis with Composer repositories of type VCS) so I can't change `"name"` to the new one because application having old name in theirs composer.lock will fail on composer install.
